### PR TITLE
Replace `context.compile_internal` by `@overload` in cmathimpl.py

### PR DESCRIPTION
**Pull request by guilhermeleobas**
~This is more of an exploratory work, and it is not ready yet.~

~The idea is to replace all occurrences of `compile_internal` by `call_overload`, introduced in this PR. Some of the motivation for this change is the ability to use `print` inside compiled function and dynamic exceptions.~

Replaces all occurrences of `context.compile_internal` by `@overload` in `cmathimpl.py`

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
`acos` test fails with the following error:

```python
numba/core/lowering.py:1392: in lower_expr
    self.incref(resty, castval)
numba/core/lowering.py:1560: in incref
    self.context.nrt.incref(self.builder, typ, val)
numba/core/runtime/context.py:382: in incref
    self._call_incref_decref(builder, typ, value, "NRT_incref")
numba/core/runtime/context.py:367: in _call_incref_decref
    meminfos = self.get_meminfos(builder, typ, value)
numba/core/runtime/context.py:355: in get_meminfos
    field = getter(val)
numba/core/datamodel/models.py:713: in getter
    raise TypeError("expecting {0} but got {1}".format(*args))
E   TypeError: expecting {float, float} but got {double, double}
```

**Comment by gmarkall** [comment]
Should this be moved from draft status now it's RFR?

**Comment by gmarkall** [comment]
gpuci run tests

**Review by gmarkall — CHANGES_REQUESTED** [review]
Many thanks for the improvements! I've left some comments on the diff.

Unfortunately this seems to have broken CUDA ufuncs e.g. - https://gpuci.gpuopenanalytics.com/job/numba/job/numba/job/prb/job/numba-prb/675/CUDA_TOOLKIT_VER=11.8,CUDA_VER=11.2,LINUX_VER=ubuntu18.04,PYTHON_VER=3.8,RAPIDS_VER=21.12/console 

Do you need some help looking into the CUDA issue, or are you happy to fix this up?

File: numba/cpython/cmathimpl.py [hunk 1]
```diff
@@ -206,13 +186,13 @@ def impl(x):
...
-@lower(cmath.sqrt, types.Complex)
-def sqrt_impl(context, builder, sig, args):
+@overload(cmath.sqrt)
+def sqrt_impl(z):
     # We risk spurious overflow for components >= FLT_MAX / (1 + sqrt(2)).
 
     SQRT2 = 1.414213562373095048801688724209698079E0
     ONE_PLUS_SQRT2 = (1. + SQRT2)
-    theargflt = sig.args[0].underlying_float
+    theargflt = z.underlying_float if isinstance(z, types.Complex) else z
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/cmathimpl.py]: Good catch! Thanks

File: numba/cpython/cmathimpl.py [hunk 2]
```diff
@@ -206,13 +186,13 @@ def impl(x):
     return impl
 
 
-@lower(cmath.sqrt, types.Complex)
-def sqrt_impl(context, builder, sig, args):
+@overload(cmath.sqrt)
+def sqrt_impl(z):
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/cpython/cmathimpl.py:186]: There's one test in parfors that call `cmath.sqrt` with an integer. I think I remember why I included the code below:

```
     theargflt = z.underlying_float if isinstance(z, types.Complex) else z 
```

Should numba cast the integer to a complex value?

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by guilhermeleobas** [comment]
The CI failure is probably related to one of the parfor tests calling `cmath.sqrt(integer)`

**Comment by guilhermeleobas** [comment]
Calling any cmath function with a non-complex value should be valid. I'll create an issue to track this and change `cmath.sqrt` to accept non-complex inputs.

**Comment by guilhermeleobas** [comment]
@gmarkall, could you run gpuci again on this PR?

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by guilhermeleobas** [comment]
I think that this PR actually introduces a regression in Numba. Numba main branch currently supports calling any cmath function with a non-complex value, and this PR would prohibit that.

**Comment by gmarkall** [comment]
gpuci run tests

**Comment by guilhermeleobas** [comment]
I don't have the bandwidth to work on this one right now. Changed the label to `in progress`.

**Pull request closed without merging**